yarn-project/noir-compiler/src/compile/noir/dependencies/dependency-manager.ts

```diff
@@ -0,0 +1,112 @@
+import { LogFn, createDebugOnlyLogger } from '@aztec/foundation/log';
+
+import { join } from 'node:path';
+
+import { NoirDependencyConfig } from '../package-config.js';
+import { NoirPackage } from '../package.js';
+import { DependencyResolver } from './dependency-resolver.js';
+
+/**
+ * Noir Dependency Resolver
+ */
+export class NoirDependencyManager {
+  #entryPoint: NoirPackage;
+  #libraries = new Map<string, NoirPackage>();
+  #dependencies = new Map<string, string[]>();
+  #log: LogFn;
+  #resolvers: readonly DependencyResolver[];
+
+  /**
+   * Creates a new dependency resolver
+   * @param resolvers - A list of dependency resolvers to use
+   * @param entryPoint - The entry point of the project
+   */
+  constructor(resolvers: readonly DependencyResolver[] = [], entryPoint: NoirPackage) {
+    this.#log = createDebugOnlyLogger('noir:dependency-resolver');
+    this.#resolvers = resolvers;
+    this.#entryPoint = entryPoint;
+  }
+
+  /**
+   * Gets dependencies for the entry point
+   */
+  public getEntrypointDependencies() {
+    return this.#dependencies.get('') ?? [];
+  }
+
+  /**
+   * A map of library dependencies
+   */
+  public getLibraryDependencies() {
+    const entries = Array.from(this.#dependencies.entries());
+    return Object.fromEntries(entries.filter(([name]) => name !== ''));
+  }
+
+  /**
+   * Resolves dependencies for a package.
+   */
+  public async resolveDependencies(): Promise<void> {
+    await this.#recursivelyResolveDependencies('', this.#entryPoint);
+  }
+
+  async #recursivelyResolveDependencies(packageName: string, noirPackage: NoirPackage): Promise<void> {
+    for (const [name, config] of Object.entries(noirPackage.getDependencies())) {
+      // TODO what happens if more than one package has the same name but different versions?
```

Comment:
I'd blow up loudly, rather than silently pick either of them. Though you'd probably need to use [semver](https://www.npmjs.com/package/semver) to collect all version restrictions and get their intersection.

Reply:
Hm, I thought about it but noir libs don't have a version number in `Nargo.toml` and because of this is makes comparing git dependencies to relative-path dependencies difficult. (e.g. contract depends on aztec and authwit via git, authwit depends on aztec via relative path)

Reply:
Should we then just use the git revision, and if we have a relative path, we "inherit" the git revision of the parent that included it? That should let us blow up if we see two different git revisions requested.